### PR TITLE
Gdrive Implementation

### PR DESCRIPTION
**Pull request by segator**
Hi,
This is my first attemp to implement gdrive using base code of the implementation of @mkhon

I modified the code of @mkhon to have the next features:

- [x] S3QL GDrive Full Implementation
- [x] Gdrive Error Handling 
- [ ]  batch requests(better performance in write/delete little files, avoid ban for too much requests)
- [x] oauth client integrated with s3ql
- [x] avoid unecesary requests
- [x] md5 checksum read/write

I don't expect you accept now the changes, I want to know I am in the good way, I want to refactor some things before merge it to main. 

About Oauth auth
I finally modified your oauth util for google storage
I added a parameter --oauth_type where you can define if you want to generate token for google storage or google drive and also I added the possibilty to use own clientID/secret.
You should modify your clientID to accept google drive, because right now you must have your own clientID to generate token.

The idea is oauth client generate  a refresh token
and when you mount s3ql you set the next values:
user: youclient_id
password:client_secret:refreshToken

Let me know what do you thing about this implementation, I'm not an expert in python so I suppose there are lot of things could be done better.

### CODE REVIEW

**Comment by Nikratio** [comment]
Thanks a lot for your contribution, and apologies for the lack of a response! I want to look at this, but I just have much less time available to spend on S3QL than in the past.

**Comment by segator** [comment]
Hi @Nikratio , I fully understand, don't worry, the implementation it's not finished yet, but I want your aprove to continue in this direction.

**Comment by szepeviktor** [comment]
Is this PR for Google Drive or http://www.g-technology.com/products/g-drive ?

**Comment by segator** [comment]
Google Drive of course

**Comment by szepeviktor** [comment]
Thanks.

**Comment by Nikratio** [comment]
Sorry for the delay! Are you still interested in finishing this? Then I will take a look.

**Comment by segator** [comment]
Yes of course, I will be happy to colaborate to your project, the implementation is pretty stable I have 3 instances withouts FS crash since 4 month ago. 
Probably some code need to be refactored to be more clear(probably my first "hard" coding with python)

The only error I don't know why is happining is sometimes when retrying uploads s3ql duplicate the file.
I add controls to avoid this but anyway is happening. So when you do FSCK you get duplicate ID errors that is easily fixing removing one of the duplicated files(are exactly equals)
Anyway should be fixed before integrate to main branch.

Some extras I add:
S3qlstat now have a prometheus exporter so you can have pretty reports, check here:
https://snapshot.raintank.io/dashboard/snapshot/74VnTwHw6s1ZNK6MZdjs6lsdo78NMxNU
s3qlstat /path --prometheus_exporter --prometheus_port

S3QL Oauth now integrate google drive  besides google storage.

**Comment by Nikratio** [comment]
Could you explain what you mean by "s3ql duplicate the file"? What fsck errros do you get specifically?

File: src/s3ql/oauth_client.py [hunk 1]
```diff
@@ -31,8 +34,14 @@ def parse_args(args):
     parser.add_debug()
     parser.add_quiet()
     parser.add_version()
+    parser.add_oauth()
 
-    return parser.parse_args(args)
+    options =  parser.parse_args(args)
+    if options.client_id == '':
```
Comment (Nikratio) [inline src/s3ql/oauth_client.py]: Can you give an example of a situation where you'd want to use a different client id?

File: src/s3ql/oauth_client.py [hunk 2]
```diff
@@ -59,20 +68,32 @@ def _parse_response(r):
...
     return r.json()
 
-def main(args=None):
 
-    if args is None:
-        args = sys.argv[1:]
 
-    options = parse_args(args)
-    setup_logging(options)
+def googleDrive(options):
```
Comment (Nikratio) [inline src/s3ql/oauth_client.py]: Please follow the conventions of the existing S3QL code (which in turn mostly follows Python PEP8). This means `CamelCase` only for classes and `under_scores` for everything else. Also, method/function names should typically be verbs so you can tell what they do. From the name, there's no way to tell what the `googleDrive` function does.

File: src/s3ql/parse_args.py [hunk 3]
```diff
@@ -182,6 +182,12 @@ def add_storage_url(self):
                           type=storage_url_type,
                           help='Storage URL of the backend that contains the file system')
 
+    def add_oauth(self):
+        self.add_argument("--oauth_type", metavar='<oauth_type>',default="google-storage",type=oauth_type,
```
Comment (Nikratio) [inline src/s3ql/parse_args.py]: I think it's better to instead require the user to pass a storage url, and infer the correct backend from that. At some point we may want to generate tokens that are specific to the given bucket (or even prefix).

File: src/s3ql/statfs.py [hunk 4]
```diff
@@ -31,6 +33,10 @@ def parse_args(args):
 
     parser.add_argument("--raw", action="store_true", default=False,
                           help="Do not pretty-print numbers")
+    parser.add_argument("--prometheus_exporter", action="store_true", default=False,
```
Comment (Nikratio) [inline src/s3ql/statfs.py]: Please move all the Prometheus stuff into a separate pull request, otherwise it's very hard to review.

File: src/s3ql/statfs.py [hunk 5]
```diff
@@ -49,32 +55,89 @@ def main(args=None):
...
-           'After de-duplication: %s (%.2f%% of total)'
-             % (pprint(dedup_size), p_dedup),
-           'After compression:    %s (%.2f%% of total, %.2f%% of de-duplicated)'
-             % (pprint(compr_size), p_compr_1, p_compr_2),
-           'Database size:        %s (uncompressed)' % pprint(db_size),
-           'Cache size:           %s, %d entries' % (pprint(cache_size), cache_cnt),
-           'Cache size (dirty):   %s, %d entries' % (pprint(dirty_size), dirty_cnt),
-           'Queued object removals: %d' % (removal_cnt,),
-           sep='\n')
+    if options.prometheus_exporter:
```
Comment (Nikratio) [inline src/s3ql/statfs.py]: I don't think that's a good idea. You should start the promethous exporter in a separate thread in mount(), and leave statfs as-is.

File: src/s3ql/backends/gdrive.py [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+import time
+import base64
+import hashlib
+import httplib2
+import ssl
+import http
+import socket
+import io
+import json
+import oauth2client.client
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Is this really required? The google storage backend uses OAuth just fine without extra modules.

File: src/s3ql/backends/gdrive.py [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+import tempfile
+
+log = logging.getLogger(__name__)
+logging.getLogger("googleapiclient.discovery").setLevel(logging.WARNING)
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Why is this needed?

File: src/s3ql/backends/gdrive.py [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+
+    MAX_PROPERTY_SIZE = 124
+    MAX_REF_CHUNK_SIZE = 100
+    VALUE_BASE64 = 'b64:'
+    VALUE_REF = 'ref:'
+    
+
+    def __init__(self, storage_url, login, password, options):
+        '''Initialize local backend
+        Login and password are ignored.
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: actually, you are using the password below

File: src/s3ql/backends/gdrive.py [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+        '''
+        # Unused argument
+        #pylint: disable=W0613
+
+        super().__init__()
+        self.login = login
+        self.password = password
+        client_secret_and_refresh_token = self.password.split(':')
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: better follow the convention of the google storage backend here

File: src/s3ql/backends/gdrive.py [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+        if len(client_secret_and_refresh_token) != 2:
+            raise AuthenticationError("Invalid Password format, must be  secretApplication:refreshToken")
+        self.client_secret = client_secret_and_refresh_token[0]
+        self.refresh_token = client_secret_and_refresh_token[1]
+        self.options = options
+        self.delete_map_state = dict()
+
+        # get google drive service
+        global credentials
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: No global variables please. What if you want to have multiple backend instances that use different credentials?

File: src/s3ql/backends/gdrive.py [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
+                if self.folder['mimeType'] != Backend.MIME_TYPE_FOLDER:
+                    raise DanglingStorageURLError(folderPath)
+            else:
+                self.folder = self.service.files().create(body={'name' : folderPath,'mimeType' : Backend.MIME_TYPE_FOLDER},fields='id').execute()
+            mainFolder=self.folder
+        else:
+            self.folder = mainFolder
+
+    def _get_credentials(self):
+        return oauth2client.client.OAuth2Credentials(access_token="1234",client_id=self.login,client_secret=self.client_secret,
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Please keep lines under 100 characters (not just here but everywhere in the code)

File: src/s3ql/backends/gdrive.py [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+            #Google best practice error says error 500 could be temporal, also We get some resumable errors with 400 Bad Request (Fully Random)
+            elif exc.resp.status >= 500 or exc.resp.status == 400:
+                return True
+        elif isinstance(exc,(BrokenPipeError,ConnectionResetError,socket.gaierror,socket.herror,socket.timeout,http.client.IncompleteRead,
+        httplib2.ServerNotFoundError,ssl.SSLZeroReturnError, ssl.SSLEOFError,ssl.SSLSyscallError,ssl.SSLError)):
+            return True
+        return False
+
+    @staticmethod
+    def _escape_string(name):
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Looks like it should be a function, why make it a class member?

File: src/s3ql/backends/gdrive.py [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+        '''Escape string'''
+        return name.replace("\\", "\\\\").replace("'", "\\'")
+
+    def _list_files(self, folder, fields, query=None):
+        '''List folder contents
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Please follow PEP8 style, i.e. one empty line after the first. Arguments should be enclosed in `*`, e.g. `the *foo* argument should be set too...`. If you want to make a list, use

```
:arg1: does this
:arg2: does that
```

(same everywhere else)

File: src/s3ql/backends/gdrive.py [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
+        Arguments:
+        folder -- Google Drive folder to list files in
+        fields -- file properties to return
+        query -- files.list query (optional)
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: I do not understand what *query* is for

File: src/s3ql/backends/gdrive.py [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+        query += "'{0}' in parents".format(folder['id'])
+        query += " and trashed = false"
+        request = self.service.files().list(q=query,pageSize=1000,fields="nextPageToken, files(%s)" % fields)
+        while request is not None:
+            results = request.execute()            
+            yield from results.get('files', [])
+            request = self.service.files().list_next(request,results)
+
+    def _lookup_file(self, name, folder=None):
+        '''Lookup file by name
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: What does this function return?

File: src/s3ql/backends/gdrive.py [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+            else:
+                return None
+
+        # nothing found
+        return None
+
+    @staticmethod
+    def _chunk_property(k, i):
+        '''Make chunk property'''
+        return 'ref({0}.{1})'.format(k, i)
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: This really doesn't justify a separate method

File: src/s3ql/backends/gdrive.py [hunk 17]
```diff
@@ -0,0 +1,599 @@
...
+
+    def _encode_metadata(self, metadata):
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: please add docstring

File: src/s3ql/backends/gdrive.py [hunk 18]
```diff
@@ -0,0 +1,599 @@
...
+                    chunks = [b64[i:i+Backend.MAX_REF_CHUNK_SIZE]
+                        for i in range(0, len(b64), Backend.MAX_REF_CHUNK_SIZE)]
+                    properties[k] = Backend.VALUE_REF + str(len(chunks))
+                    for i, chunk in enumerate(chunks):
+                        properties[self._chunk_property(k, i)] = chunk
+            else:
+                properties[k] = v
+        return properties
+
+    def _decode_metadata(self, f):
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: This looks like you really should be using `thaw_basic_mapping()` and `free_basic_mapping` instead (from `s3ql.common`)

File: src/s3ql/backends/gdrive.py [hunk 19]
```diff
@@ -0,0 +1,599 @@
...
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)  
+        return ObjectR(self.service, f, self._decode_metadata(f))
+
+    @prepend_ancestor_docstring
+    def open_write(self, key, metadata=None, is_compressed=False):
+        log.debug("key: {0}".format(key))
+        if metadata is None:
+            metadata = dict()
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: If metadata is never `None`, why does _encode_metadata check for it?

File: src/s3ql/backends/gdrive.py [hunk 20]
```diff
@@ -0,0 +1,599 @@
...
+        }
+        log.debug("metadata: {0}, body: {1}".format(metadata, body))
+        return ObjectW(self.service,self, key,body)
+
+    @prepend_ancestor_docstring
+    def clear(self):
+        log.debug("")
+        for f in self._list_files(self.folder, "id"):
+            self._delete_by_id(f['id'])            
+    '''
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: What's happening here? Why is this all commented out?

File: src/s3ql/backends/gdrive.py [hunk 21]
```diff
@@ -0,0 +1,599 @@
...
+    @copy_ancestor_docstring
+    def delete(self, key, force=False, is_retry=False):
+        log.debug("key: {0}".format(key))
+        query = "name = '{0}'".format(self._escape_string(key))
+        found = False
+        for f in self._list_files(self.folder, "id", query):
+            found = True
+            self._delete_by_id(f['id'])
+        if not found:
+            if force or is_retry:
```
Comment (Nikratio) [inline src/s3ql/backends/gdrive.py]: Why ignore the error on retry?

File: setup.py [hunk 22]
```diff
@@ -134,7 +134,9 @@ def main():
                      'requests',
                      'defusedxml',
                      'dugong >= 3.4, < 4.0',
-                     'llfuse >= 1.0, < 2.0' ]
+                     'llfuse >= 1.0, < 2.0',
+                     'google-api-python-client >= 1.4.2']
```
[the remark this answers is not in the document]
Comment (segator) [reply setup.py]: I suppose you mean google-api-python-client, it's the official google drive SDK, yes of course is necesary unless you want to reinvent the wheel and implement all the API request with your own http library

File: src/s3ql/oauth_client.py [hunk 23]
```diff
@@ -8,6 +8,9 @@
 
 from .logging import logging, setup_logging, QuietError
 from .parse_args import ArgumentParser
+from oauth2client.client import OAuth2WebServerFlow
+from oauth2client.client import AccessTokenCredentials
+from  oauth2client.client import OOB_CALLBACK_URN
```
[the remark this answers is not in the document]
Comment (segator) [reply src/s3ql/oauth_client.py]: I suppose could be implemented with the existing oauth but I wanted use official things.

File: src/s3ql/oauth_client.py (hunk 1 above)
Comment (segator) [reply src/s3ql/oauth_client.py]: google drive limit max number of request per minut/day depends of the  application, so maybe you could have an google application with a higher cap.(This is my case)

File: src/s3ql/oauth_client.py (hunk 2 above)
Comment (segator) [reply src/s3ql/oauth_client.py]: You are right, pending to refactor

File: src/s3ql/parse_args.py (hunk 3 above)
Comment (segator) [reply src/s3ql/parse_args.py]: I don't know if I understand, what you checked in the code is the parameter for choice oauth_type storage/drive when generating user/pass for your specific google drive/storage account.

File: src/s3ql/statfs.py (hunk 4 above)
Comment (segator) [reply src/s3ql/statfs.py]: Okey, I will create another branch and move prometheus things to other PR to check after work with gdrive implementation

File: src/s3ql/statfs.py (hunk 5 above)
Comment (segator) [reply src/s3ql/statfs.py]: Yes you are right I tried but I am not an expert In python I getted the FS fully blocked, maybe you could help me with that.

File: src/s3ql/backends/gdrive.py (hunk 6 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: this imports are because the function is_temp_failure to check if the error is one of them. We should wrapp all the errors in a generic Exception probably

File: src/s3ql/backends/gdrive.py (hunk 7 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: Because an error of google SDK all the request were shown on the log, impossible to read anything because to much information, this fix the issue avoiding all the requests are writed in the s3ql log

File: src/s3ql/backends/gdrive.py (hunk 8 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: Yes, I miss to remove this :P)

File: src/s3ql/backends/gdrive.py (hunk 10 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: I wanted to avoid every thread refresh google refresh token by itself so this way the token only needed to refresh on one of them and all remaining can get the refreshed token. any idea how to do it better?

File: src/s3ql/backends/gdrive.py (hunk 14 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: google drive have their own "SQL" to find files so this var allow you to use  gdrive query language  to find files.

File: src/s3ql/backends/gdrive.py (hunk 15 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: It returns None in case nothing found or an array of properties (Id, name,size..) of the file found

File: src/s3ql/backends/gdrive.py (hunk 16 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: to be refactored

File: src/s3ql/backends/gdrive.py (hunk 17 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: right

File: src/s3ql/backends/gdrive.py (hunk 18 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: I will check it, I didn't know about this

File: src/s3ql/backends/gdrive.py (hunk 19 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: are you fully right never is None? if you are, could be removed

File: src/s3ql/backends/gdrive.py (hunk 20 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: it was an attemt to implement delete_multi, but it is very hard because gdrive no process all of them as block, most of them could fail, what do you do in this case? rexecute until 0 pending/error requests? probably you will get a ban for too much requests.

File: src/s3ql/backends/gdrive.py (hunk 21 above)
Comment (segator) [reply src/s3ql/backends/gdrive.py]: gdrive not always works "in sync" , some times you will get an error deleting a file but in the next retry request you find out the file doesn't exist anymore, it's weird, I add this condition to avoid s3ql crash when this kind of things happen.
When google scale up/down their infrastructure is common get error 500(it's in their documentation) I guess when this happens you could have an error 500 but the request is procesed correctly.

File: setup.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Nikratio) [reply setup.py]: The problem with these official SKDs is that often they are designed to do much more than what an application like S3QL wants a library to do. For example, S3QL needs control over threading, connecting, re-trying requests etc. If the SDK does all these things automatically, that's super convenient for quick experimentation, but S3QL would not work very well. That's the reason why S3QL isn't using the official Amazon S3 or Google Storage SDKs either.

File: src/s3ql/oauth_client.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (Nikratio) [reply src/s3ql/oauth_client.py]: I think in that case I'd rather switch to a different flow for both Google Storage and Google Drive.  Or is there a good reason to use different flows?

File: src/s3ql/parse_args.py (hunk 3 above)
Comment (Nikratio) [reply src/s3ql/parse_args.py]: Yes. Drop that parameter, and require the user to specify the storage URL. Then you can look at the prefix (`gs://` or `gdrive://`) to determine whether the token is for Google Drive or Google Storage.

File: src/s3ql/statfs.py (hunk 5 above)
Comment (Nikratio) [reply src/s3ql/statfs.py]: Try to model your implementation after the `CommitThread` in `mount.py`, that should work fine. I'll help further in the other pull request.

File: src/s3ql/backends/gdrive.py (hunk 6 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: I'm talking specifically about `oauth2client.client`. Just like with `google-api-python-client` I am worried that it may be doing much more than we want. To clarify: if omitting this dependency means hundreds of lines of extra code we should of course use the module. But if a from scratch implementation of the subset of functions that we require for S3QL is just a hundred lines or so, then I wouldn't want to use an external module for that.

File: src/s3ql/backends/gdrive.py (hunk 7 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: And this hasn't been fixed? That's a good example of reasons to avoid the library then :-). If it has been fixed, we should explicitly depend on a fixed version of the module (if we have to use the module at all).

File: src/s3ql/backends/gdrive.py (hunk 10 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: Yes. Take a look at how the Google Storage backend does it (look for `_refresh_lock`). I think one of the swift backends also has similar functionality.

File: src/s3ql/backends/gdrive.py (hunk 14 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: Could you please put something along these lines in the docstring? A link to the webpage that describes the query language would be nice too.

File: src/s3ql/backends/gdrive.py (hunk 15 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: Please put that in the docstring then :-). But don't use the word array. I am quite sure that what you're returning is not a Python array (cf https://docs.python.org/3/library/array.html)

File: src/s3ql/backends/gdrive.py (hunk 19 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: No, what I'm saying is that here you ensure that it *metadata* will never be `None` from here on. But then you call another function (sorry, can't see the context at the moment, may be something like `_encode_metadata`) that checks if *metadata* is None and in that case returns `None` - defeating the check above.

File: src/s3ql/backends/gdrive.py (hunk 20 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: I do not understand what you mean with *because gdrive no process all of them as block, most of them could fail*. If there is no way to implement `delete_multi`, just don't implement it. The higher layers will fail back to repeated `delete` calls instead.

I expect that the server will send you warnings to slow down before banning you completely. If the method raises the right exceptions, S3QL will honor these warnings and everything should work fine. If the *google-api-python-client* module doesn't give you access to these early warnings then... we shouldn't use it.

File: src/s3ql/backends/gdrive.py (hunk 21 above)
Comment (Nikratio) [reply src/s3ql/backends/gdrive.py]: Alright. Could you please put this into the code as a comment?

**Comment by Nikratio** [comment]
Friendly ping. Do you think you'll have time to resolve the open issues in the near future? If not, I'll close this pull request for now.

**Comment by Nikratio** [comment]
Thanks for the quick reply! Are you referring to this comment?

> I expect that the server will send you warnings to slow down before banning you completely. If the method raises the right exceptions, S3QL will honor these warnings and everything should work fine. If the google-api-python-client module doesn't give you access to these early warnings then... we shouldn't use it.

If so, then it's not clear to me why this means that you can't use the official sdk. Why can't you simply not implement `delete_multi` as I suggested?

If you refer to the `logging` related comment: can't you depend on a fixed version of the SDK?

**Comment by Nikratio** [comment]
You misunderstood. You can use the official SDK as long as you can solve the issues described in the comments.

**Comment by segator** [comment]
ok sorry, I will check all the comments(hopefully this month) again and try to fix all of them, gdrive work pretty stable and fast with s3ql, I have a s3ql mountpoint running more than 1 year and a half without issues, and other has been upgraded to have persistent cache and works pretty nice too.
Thanks for your work!

**Comment by Nikratio** [comment]
Closing this for now as discussed above. Please do reopen once the open issues are resolved!

**Comment by eleaner** [comment]
Guys, in the name of less cash-heavy users of S3QL, would you be able to revive the discussion?

I guess al ot of us here would love to see it running

**Comment by stickenhoffen** [comment]
Yeah I agree; Google Drive would be an amazing addition.

**Comment by eleaner** [comment]
@stickenhoffen 
sadly not much interest here
